File: AudioKit/Operations/Signal Input and Output/AKAudioOutput.m
//
//  AKAudioOutput.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 6/9/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "AKAudioOutput.h"

@implementation AKAudioOutput
{
    AKStereoAudio *aStereoOutput;
}

- (instancetype)initWithInput:(AKParameter *)source
{
    return [self initWithLeftAudio:source rightAudio:source];
}

- (instancetype)initWithAudioSource:(AKParameter *)audioSource
{
    return [self initWithLeftAudio:audioSource rightAudio:audioSource];
}

- (instancetype)initWithStereoAudioSource:(AKStereoAudio *)stereoAudio {
    self = [super initWithString:[self operationName]];
    if (self) {
        aStereoOutput = stereoAudio;
        self.state = @"connectable";
        self.dependencies = @[stereoAudio];
    }
    return self;
}

- (instancetype)initWithLeftAudio:(AKParameter *)leftAudio
                       rightAudio:(AKParameter *)rightAudio
{
    self = [super initWithString:[self operationName]];
    if (self) {
        aStereoOutput = [[AKStereoAudio alloc] initWithLeftAudio:leftAudio
                                                      rightAudio:rightAudio];
        self.state = @"connectable";
        self.dependencies = @[leftAudio, rightAudio];
    }
    return self; 
}

// Csound prototype: outs asig1, asig2
- (NSString *)stringForCSD {
    return [NSString stringWithFormat:@"outs %@", aStereoOutput];
}

@end

File: AudioKit/Operations/Signal Input and Output/AKAudioOutput.swift
//
//  AKAudioOutput.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 9/10/15.
//  Copyright © 2015 AudioKit. All rights reserved.
//

import Foundation

/** Output audio to the speakers */
@objc class AKAudioOutput : AKParameter {
    
    private var input = AKParameter()

    /** Create an instance of an audio output
    - parameter input: The audio signal
    */
    init(input sourceInput: AKParameter) {
        super.init()
        input = sourceInput
        dependencies = [input]
    }
    
    /** Place the audio signal in the output chain
    */
    override func compute() {
        sp_out(AKManager.sharedManager.data, 0, input.leftOutput)
        sp_out(AKManager.sharedManager.data, 1, input.rightOutput)
    }
    
}
